### PR TITLE
Users achievements

### LINKED ISSUE #1: Achievements
As a Workshop admin I want to add the Achievements to users profile so that all users will be able to know about the level of users of the service provider 

**Role**
Workshop admin 

**Precondition**
Service providers are created, workshops are created, parents applied their children to the workshops,  Workshop admin wants to add the Achievements to the user's profile so that all users will be able to know about the level of users of the service provider 

### Acceptance Criteria
SCENARIO #1 Workshop admin adds the Achievement to the users profile 
 
GIVEN Workshop admin enters the cabinet
AND opens tab “Гуртки” 

AND opens needed Workshop 
AND selects tab “Здобувачі”
AND selects the user that wants to add the achievement
AND users info-page opens
Здобувачі 5

2021-12-04 18_47_49-MoE To Do's docx

AND clicks edit button “Редагувати”
AND edit page opens

AND selects the achievement from the dropdown list
AND inputs date
AND inputs the name of the achievement 
AND inputs name of the Coach
WHEN clicks “Зберегти”
THEN achievement is saved
AND all users can view new achievements on the tab “Про гурток” at the chapter “Досягнення
AND all users can view all achievements on the tab “Про заклад” at the chapter “Досягнення

Досягнення 9999

**Business rules**
All current achievements are displayed in the current workshop and also all achievements are displayed at the service providers tab “Про заклад”
 
 
Dropdown-fields of achievements: 
 
1)Переможці міжнародних та всеукраїнських спортивних змагань (індивідуальних та командних)
2)Призери та учасники міжнародних, всеукраїнських та призери регіональних конкурсів і виставок наукових, технічних, дослідницьких, інноваційних, ІТ проектів
3)Реципієнти міжнародних грантів
4)Призери міжнародних культурних конкурсів та фестивалів
5)Соціально активні категорії учнів
6)Цифрові інструменти Google для закладів вищої та фахової передвищої освіти
7)Переможці та учасники олімпіад міжнародного та всеукраїнського рівнів
 
List:
https://docs.google.com/spreadsheets/d/1yD1PH5b-4RjtNjNkTZYhOERCRvD3aFK71KVL2noImAw/edit#gid=813362464 
 
Information is entered by the admin 
Parents can only view
Information is visible for viewing to all admins

**Epic link**
E.g.: Epic #100 [Epic](https://github.com/ita-social-projects/OoS-Backend/labels/100)

**Labels to be added**
"User story", Correspondind Functionality (e.g. User), Priority ("pri: "), "Corresponding Epic"

### Tasks 
1. - [ ] sample task.

### PR DESCRIPTION
**Pull request by Tfront1**
Handy data for manual testing from an existing database dump:

For creating new achieve:

Group administrator
Login: rulmacagna@vusra.com
Password: zi__U-6F

applicationId:
08dbbb7b-31b3-4244-87a1-7cfef0cb995b

trainerId:
08da8478-2160-4e69-8fcb-8313b49a2e60

-----------------------

Related data with previous ones:

childId:
08db035d-c372-4745-8f8c-fc6959de196c

workshopId:
08da8478-215e-41e8-8813-e4f25e3a208b

parentId:
08da847c-2f6d-4327-8184-b1a11c8f7008

providerId:
08da842d-12fc-4865-85c5-ec6e6142abad

### CODE REVIEW

File: OutOfSchool/OutOfSchool.DataAccess/Models/ChildAchievement.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+{
+    public Guid Id { get; set; }
+
+    public int ChildAchievementTypeId { get; set; }
+
+    public DateTime Date { get; set; }
+
+    public string Name { get; set; }
+
+    public Guid TrainerId { get; set; }
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.DataAccess/Models/ChildAchievement.cs:14]: If you're using Trainer(Teacher) Id then you probably don't need Workshop Id as it can be traced through teacher.

File: OutOfSchool/OutOfSchool.DataAccess/Repository/ApplicationRepository.cs [hunk 2]
```diff
@@ -83,4 +84,11 @@ public async Task<int> UpdateAllApprovedApplications()
                 (int)ApplicationStatus.Approved)
             .ConfigureAwait(false);
     }
+
+    public async Task<IEnumerable<Application>> GetForWorkshopChild(Guid childId, Guid workshopId)
+    {
+        return await dbContext.Applications
+            .Where(x => x.ChildId == childId && x.WorkshopId == workshopId)
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.DataAccess/Repository/ApplicationRepository.cs:91]: In the logic of the application you can have applications to the workshop that can be in neutral (pending), negative (denied, etc.), positive (accepted, studying) status. For this story you need to somehow make sure that the child is actually accepted/studying in this workshop.

File: OutOfSchool/OutOfSchool.WebApi/Services/ChildAchievementService.cs [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    }
+
+    public async Task<ChildAchievementCreationResponseDto> CreateAchievement(ChildAchievementCreationRequestDto childAchievementCreationRequestDto, string userId)
+    {
+        _ = childAchievementCreationRequestDto ?? throw new ArgumentNullException(nameof(childAchievementCreationRequestDto));
+
+        logger.LogDebug(
+            $"Started creation of a new child achievement {nameof(childAchievementCreationRequestDto)}:{childAchievementCreationRequestDto}.");
+
+        var type = (await childAchievementTypeRepository.GetById(childAchievementCreationRequestDto.ChildAchievementTypeId).ConfigureAwait(false))
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.WebApi/Services/ChildAchievementService.cs]: it does make your controllers simpler by outsourcing some response generating logic to exception handler. But exceptions are not flow control and you should not use it as such. It's one of those bad practices that allow you to write code faster, but come back as new headache later (lazy loading is another example of this) so we're trying to refactor our code with this in mind. Exceptions are for situations when your code can't continue it's normal operations or you need to forcefully stop it. If for some reason something is not found in database - it not an exception, it's life :).
We have a `OperationResult` class somewhere :) and it's better to use it as a return type if you know there might be minor problems. In controllers you can check if it's success or not and act accordingly.

**Review by DmyMi — CHANGES_REQUESTED** [review]

**Review by DmyMi — APPROVED** [review]
